File: examples/cpp/core/serialization/serialization.cpp
#include "serialization_reflection.h"

#include <bond/core/bond.h>
#include <bond/stream/output_buffer.h>

using namespace examples::serialization;


int main()
{
    Struct obj;

    obj.n = 0x1000;
    obj.str = "test";
    obj.items.push_back(3.14);
    obj.items.push_back(0);

    // bond::OutputBuffer implements the Bond output stream interface on top of
    // a series of memory buffers.
    bond::OutputBuffer output;

    // Use the Compact Binary protocol for encoding; the data will be written to
    // the OutputBuffer which will allocate memory as needed.
    bond::CompactBinaryWriter<bond::OutputBuffer> writer(output);

    // Serialize the object
    Serialize(obj, writer);

    // At this point the OutputBuffer contains a serialized representation of
    // the object. Depending on the size, the serialized data may be spread
    // across one or more memory blocks. Applications can get all these blocks
    // using the GetBuffers method, or merge them into a single block using
    // GetBuffer method.
    bond::blob data = output.GetBuffer();

    // Use the Compact Binary protocol for decoding (the same protocol that was
    // used for encoding) and InputBuffer which implements the input stream
    // interface on top of a memory blob.
    bond::CompactBinaryReader<bond::InputBuffer> reader(data);

    // De-serialize the object
    Struct obj2;
    Deserialize(reader, obj2);

    return 0;
}

File: examples/java/core/serialization/src/main/java/com/microsoft/bond/examples/Serialization.java
package com.microsoft.bond.examples;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;

import com.microsoft.bond.Deserializer;
import com.microsoft.bond.Serializer;

import com.microsoft.bond.protocol.CompactBinaryReader;
import com.microsoft.bond.protocol.CompactBinaryWriter;

import com.microsoft.bond.examples.serialization.Struct;

public class Serialization {

    public static void main(final String[] args) throws IOException {

        final Struct obj = new Struct();
        obj.n = 0x1000;
        obj.str = "test";
        obj.items.add(3.14D);
        obj.items.add(0D);

        final ByteArrayOutputStream output = new ByteArrayOutputStream();
        final CompactBinaryWriter writer = new CompactBinaryWriter(output, (short) 1);

        final Serializer<Struct> serializer = new Serializer<>();
        serializer.serialize(obj, writer);

        final ByteArrayInputStream input = new ByteArrayInputStream(output.toByteArray());

        final CompactBinaryReader reader = new CompactBinaryReader(input, (short) 1);
        final Deserializer<Struct> deserializer = new Deserializer<>(Struct.BOND_TYPE);
        final Struct obj2 = deserializer.deserialize(reader);

        assert obj.equals(obj2) : "Roundtrip failed";
    }

}
